src/nvim/ex_getln.c

```diff
@@ -2263,20 +2280,56 @@ static void draw_cmdline(int start, int len)
     msg_outtrans_len(ccline.cmdbuff + start, len);
 }
 
+void ui_ext_cmdline_char(int c, int shift)
```

Comment:
Maybe my point is better illustrated by an example:

```
/// Undisplay the popup menu (later).
void pum_undisplay(void)
{
  pum_is_visible = false;
  pum_array = NULL;

  if (pum_external) {
    Array args = ARRAY_DICT_INIT;
    ui_event("popupmenu_hide", args);
  } else {
    redraw_all_later(SOME_VALID);
    redraw_tabline = true;
    status_redraw_all();
  }
}
```
The relevant abstraction here is already `pum_undisplay()`. If we add yet another layer, the actual `ui_event` (or maybe later `ui_popupmenu_hide() ` auto generated wrapper), will be doubly divorced from the true source of the ui change, the calls to `pum_undisplay()`.  In both this and the `cmdline_putchar` case that will be a very short function one is forced to analyze separately (from the true abstraction point) to make sure it don't have extra conditions etc. On the other side, `ui_ext_tabline_show` and `cmdline_show` makes sense as there actually are some logic going on. Also, it would make sense to extract the external logic  in `pum_display` to _one_ function, even if it might result in two different events. To put it bluntly, I don't see how a bad pattern becomes a good pattern just because one uses it consistently.